stencila/pyla/servers.py

```diff
@@ -0,0 +1,289 @@
+        super().__init__(message)
+        self.code = code
+        self.message = message
+        self.data = data
+
+
+class StreamServer:
+    """
+    A server that communicates using length-prefixed JSON-RPC messages over streams or sockets.
+
+    Python implementation of Executa's
+    [StreamServer](https://github.com/stencila/executa/blob/v1.0.0/src/stdio/StreamServer.ts#L10)
+    """
+
+    interpreter: Interpreter
+    input_stream: StreamType
+    output_stream: StreamType
+
+    def __init__(self, interpreter: Interpreter, input_stream: StreamType, output_stream: StreamType) -> None:
+        self.interpreter = interpreter
+        self.input_stream = input_stream
+        self.output_stream = output_stream
+
+    def read_message(self) -> typing.Iterable[str]:
+        """Read a length-prefixed message from the input stream then repeat."""
+        while True:
+            yield message_read(self.input_stream)
+
+    def write_message(self, message: str) -> None:
+        """Write a length-prefixed message to the output stream."""
+        message_write(self.output_stream, message)
+
+    def receive_message(self, message: str) -> str:
+        """
+        Receive a JSON-RPC request and send back a JSON-RPC response.
+
+        The response may have a JSON-RPC `error` if the request was bad.
+
+        Python implementation of Executa's
+        [Server.receive](https://github.com/stencila/executa/blob/v1.0.1/src/base/Server.ts#L61).
+        """
+        request_id = None
+        result = None
+        error = None
+
+        try:
+            try:
+                request = json.loads(message)
+            except BaseException as exc:
+                raise JsonRpcError(JsonRpcErrorCode.ParseError, 'Parse error: %s' % repr(exc))
+
+            request_id = request.get('id')
+            method = request.get('method')
+            params = request.get('params')
+
+            if method == 'manifest':
+                result = {
+                    # Temporary. Should return real manifest
+                    'capabilities': {
+                        'manifest': True,
+                        'execute': True
+                    }
+                }
+            elif method == 'execute':
+                node = params.get('node')
+                if node is None:
+                    raise JsonRpcError(JsonRpcErrorCode.InvalidParams, 'Invalid params: "node" is missing')
+
+                code = from_dict(node)
+                if isinstance(code, CodeChunk):
+                    to_execute = simple_code_chunk_parse(code)
+                else:
+                    to_execute = code
+                self.interpreter.execute([to_execute], {})
+
+                result = code
+            else:
+                raise JsonRpcError(JsonRpcErrorCode.MethodNotFound, 'Method not found: %s' % method)
+
+        # pylint: disable=broad-except
+        except BaseException as exc:
```

Comment:
I think this would be more pythonic with multiple exception handlers

```python
except JsonRpcError as exc:
    error = exc
except Exception as exc:
    logging.exception()
    error = JsonRpcError(JsonRpcErrorCode.ServerError, 'Internal error: {}'.format(exc))

Reply:
Yep, good point, that was a lazy porting of JS logic